api/storage_service.cc

```diff
@@ -600,6 +602,161 @@ rest_get_token_endpoint(http_context& ctx, sharded<service::storage_service>& ss
+                        if (v.HasMember(name)) {
+                            if (!v[name].IsInt()) {
+                                throw bad_param_exception(fmt::format("'{}' must be an integer", name));
+                            }
+                            port_map[name] = v[name].GetInt();
+                        }
+                    }
+                    if (port_map.empty()) {
+                        throw bad_param_exception("At least one port field ('port', 'tls_port', 'alternator_port', 'alternator_https_port') must be specified");
+                    }
+
+                    db::system_keyspace::client_route route{
+                        {v["connection_id"].GetString(), v["connection_id"].GetStringLength()},
+                        utils::UUID{v["host_id"].GetString()}, // Valid UUID cannot have null terminator in the middle
+                        {v["address"].GetString(), v["address"].GetStringLength()},
+                        port_map["port"],
+                        port_map["tls_port"],
+                        port_map["alternator_port"],
+                        port_map["alternator_https_port"]
+                    };
+
+                    auto mut = co_await group0_client.sys_ks().make_update_client_route_mutation(batch.write_timestamp(), route);
+                    batch.add_mutation(std::move(mut), "update_client_route");
+                }
+                co_await std::move(batch).commit(group0_client, ss.get_abort_source(), {});
+            } catch (const ::service::group0_concurrent_modification& ex) {
+                apilog.warn("Failed to set client routes due to guard conflict, retries={}", retries);
+                if (retries--) {
+                    continue;
+                }
+                throw;
+            }
+            break;
+        }
+    });
+    co_return seastar::json::json_void();
+}
+
+static
+future<json::json_return_type>
+rest_delete_client_routes(http_context& ctx, sharded<service::storage_service>& ss, service::raft_group0_client& group0_client, std::unique_ptr<http::request> req) {
+    if (!ss.local().get_feature_service().client_routes) {
+        apilog.warn("delete_client_routes: called before the cluster feature was enabled");
+        throw std::runtime_error("delete_client_routes requires all nodes to support the CLIENT_ROUTES cluster feature");
+    }
+
+    rapidjson::Document root;
+    auto content = co_await util::read_entire_stream_contiguous(*req->content_stream);
+    root.Parse(content.c_str());
+    if (!root.IsArray()) {
+        throw bad_param_exception("Body must be a JSON array");
+    }
+
+    co_await ss.invoke_on(0, [&group0_client, &root] (service::storage_service& ss) -> future<> {
+        int retries = 10;
+        while (true) {
+            try {
+                auto batch = service::group0_batch(co_await group0_client.start_operation(ss.get_abort_source()));
```

Comment:
I wonder what will happen if we try to update/delete a huge number of rows in one of those API endpoints. I see
```cpp
// If size or number of mutations is too big for raft too handle it will be
// rejected.
class group0_batch {
```
but I'm not sure what "rejected" means. Hopefully, we will get an exception (which one?) here, and this command won't be applied as a no-op unnoticed.